proposals/2024-01-26_classic-histograms-stored-as-native-histograms.md

```diff
@@ -0,0 +1,177 @@
+  * In the Prometheus/OpenMetrics ProtoBuf [format](https://github.com/prometheus/client_model/blob/d56cd794bca9543da8cc93e95432cd64d5f99635/io/prometheus/client/metrics.proto#L115-L122) float and integer numbers are explicitly transferred in different fields.
+
+### Writing/storage
+
+* Remote write protocol to support transferring custom bucket definitions as part of the current Histogram DTO.
+* TSDB should reuse the existing chunk types.
+  * If the sample has exponential schema and buckets, store the exponential schema and buckets as currently.
+  * Otherwise if the sample has custom bucket definition and buckets, use a special schema number and store the custom bucket definitions at the beginning of the chunk once and then the bucket values as before. In the rare case when the bucket layout changes, start a new chunk to only store the custom bucket definitions once per chunk.
+
+### Reading via PromQL
+
+* Direct series selection of `metric{}` would look up `metric{}` and return the exponential histograms or custom histograms as found in the chunks.
+
+  > Note: the [exposition format](https://prometheus.io/docs/prometheus/latest/querying/api/#native-histograms) of histograms in PromQL is already forward compatible with custom buckets as the data contains the boundaries and inclusion settings.
+* Regex series selector of `__name__=~”metric.*”` would return `metric{}` and revert to the direct case.
+* Direct series selection of `metric_bucket{}`, `metric_count{}` and `metric_sum{}` would return nothing as these won’t exist.
+* The function histogram_quantile, histogram_fraction, histogram_stddev, histogram_stdvar would now potentially select custom bucket and exponential histogram samples at the same time.
+  * This is a pure math problem to find a good approximation when this happens. Will be worked out in the implementation.
+* The function histogram_count would now potentially select custom bucket and exponential histogram samples at the same time. The interpretation of count is the same for custom and exponential histograms so this is not a problem.
+* The function histogram_sum would now potentially select custom bucket and exponential histogram samples at the same time. The interpretation of count is the same for custom and exponential histograms so this is not a problem.
+
+### Migration path for users
+
+Related CNCF slack [thread](https://cloud-native.slack.com/archives/C02KR205UMU/p1706087513992019?thread_ts=1706019629.852979&cid=C02KR205UMU).
+
+Use case: neither custom histograms nor exponential histograms in use (i.e. both features are off).
+
+* Enable feature to store custom bucket native histograms, scrape them in parallel to classic (via `scrape_classic_histograms` option), and do nothing else for a while (i.e. production usage still uses classic, change no dashboards and alerts.
```

Comment:
YES those are the options we a proposing currently, although we have not talked about the granularity of control? Is this a feature flag? A scrape level flag? A metric level flag?

> > Once you are done, switch off scraping classic histograms (or drop via metric relabel as needed).
> 
> Ah... hm. This is bit prone to mistakes though. Ideally option to convert to custom have configuration to delete classics automatically 🤔 We can decide on this later on (implementation detail a bit)

I don't understand this sentence, can you give an example?

Reply:
I mean we want in future to have people drop classic histograms always, right? So dropping classic metrics will be a default. If for this you need manual relabelling for this (prone to error as -- making sure to not accidentally drop non classic histogram metrics and include all with _sum and _count etc), for many it will be simply to difficult to achieve. I think we need dedicated mode for that "removal". WDYT?

Reply:
Yes, the proposal is straightforward and drops all classic metrics by default. It doesn't depend on naming/relabelling, it's looking at the type, see [link](https://github.com/prometheus/prometheus/blob/fb552d290d8935c196a724a49d763080f3d7ff3a/model/textparse/protobufparse.go#L179). 

Conversely you can enable scraping all classic metrics with a single option.

However in reality these migration take a long time, so if you enabled custom histograms and scrape classic series at the same time, you might want to start dropping some classic series (esp. if they are high cardinality) and in that case you can add a relabel rule to drop some `<metric>_(bucket|count|sum)`. We don't actually have to do anything other then mention in documentation , since this is an existing feature.

Reply:
Thanks! makes sense